File: packages/shared/src/catalog.js
export const MODEL_CATALOG = [
  { id: 'copilot:gpt-4.1',           provider: 'copilot',    model: 'gpt-4.1',                   label: 'GPT-4.1 (Copilot)',           tier: 'low',       cost: 'usage-based' },
  { id: 'copilot:gpt-5-mini',        provider: 'copilot',    model: 'gpt-5-mini',                label: 'GPT-5 mini (Copilot)',        tier: 'low',       cost: 'usage-based' },
  { id: 'copilot:gpt-5.4-mini',      provider: 'copilot',    model: 'gpt-5.4-mini',              label: 'GPT-5.4 mini (Copilot)',      tier: 'medium',    cost: 'usage-based' },
  { id: 'copilot:gpt-5.4',           provider: 'copilot',    model: 'gpt-5.4',                   label: 'GPT-5.4 (Copilot)',           tier: 'high',      cost: 'usage-based' },
  { id: 'copilot:claude-haiku-4.5',  provider: 'copilot',    model: 'claude-haiku-4.5',          label: 'Claude Haiku 4.5 (Copilot)',  tier: 'low',       cost: 'usage-based' },
  { id: 'copilot:claude-sonnet-4.6', provider: 'copilot',    model: 'claude-sonnet-4.6',         label: 'Claude Sonnet 4.6 (Copilot)', tier: 'high',      cost: 'usage-based' },
  { id: 'claude-cli:haiku',          provider: 'claude-cli', model: 'haiku',                     label: 'Claude Haiku (CLI)',          tier: 'low',       cost: 'token-billed' },
  { id: 'claude-cli:sonnet',         provider: 'claude-cli', model: 'sonnet',                    label: 'Claude Sonnet (CLI)',         tier: 'high',      cost: 'token-billed' },
  { id: 'claude-cli:opus',           provider: 'claude-cli', model: 'opus',                      label: 'Claude Opus (CLI)',           tier: 'reasoning', cost: 'token-billed' },
  { id: 'sdk:haiku',                 provider: 'sdk',        model: 'claude-haiku-4-5-20251001', label: 'Claude Haiku (SDK)',          tier: 'low',       cost: 'token-billed' },
  { id: 'gemini:2.5-flash',          provider: 'gemini',     model: 'gemini-2.5-flash',          label: 'Gemini 2.5 Flash',            tier: 'low',       cost: 'token-billed' },
  { id: 'opencode:big-pickle',              provider: 'opencode',     model: 'big-pickle',              label: 'Big Pickle (opencode)',              tier: 'low',       cost: 'free (0x)' },
  { id: 'opencode:deepseek-v4-flash-free',  provider: 'opencode',     model: 'deepseek-v4-flash-free',  label: 'DeepSeek V4 Flash Free (opencode)',  tier: 'low',       cost: 'free (0x)' },
  { id: 'opencode-go:deepseek-v4-flash',    provider: 'opencode-go',  model: 'deepseek-v4-flash',       label: 'DeepSeek V4 Flash (opencode Go)',    tier: 'low',       cost: 'subscription' },
  { id: 'opencode-go:mimo-v2.5',            provider: 'opencode-go',  model: 'mimo-v2.5',               label: 'Mimo v2.5 (opencode Go)',            tier: 'low',       cost: 'subscription' },
  { id: 'opencode-go:minimax-m2.5',         provider: 'opencode-go',  model: 'minimax-m2.5',            label: 'MiniMax M2.5 (opencode Go)',         tier: 'low',       cost: 'subscription' },
  { id: 'opencode-go:glm-5',               provider: 'opencode-go',  model: 'glm-5',                   label: 'GLM-5 (opencode Go)',                tier: 'low',       cost: 'subscription' },
  { id: 'opencode-go:glm-5.1',             provider: 'opencode-go',  model: 'glm-5.1',                 label: 'GLM-5.1 (opencode Go)',              tier: 'low',       cost: 'subscription' },
  { id: 'opencode-go:mimo-v2.5-pro',       provider: 'opencode-go',  model: 'mimo-v2.5-pro',           label: 'Mimo v2.5 Pro (opencode Go)',        tier: 'medium',    cost: 'subscription' },
  { id: 'opencode-go:minimax-m2.7',        provider: 'opencode-go',  model: 'minimax-m2.7',            label: 'MiniMax M2.7 (opencode Go)',         tier: 'medium',    cost: 'subscription' },
  { id: 'opencode-go:qwen3.6-plus',        provider: 'opencode-go',  model: 'qwen3.6-plus',            label: 'Qwen 3.6 Plus (opencode Go)',        tier: 'medium',    cost: 'subscription' },
  { id: 'opencode-go:deepseek-v4-pro',     provider: 'opencode-go',  model: 'deepseek-v4-pro',         label: 'DeepSeek V4 Pro (opencode Go)',      tier: 'high',      cost: 'subscription' },
  { id: 'opencode-go:minimax-m3',          provider: 'opencode-go',  model: 'minimax-m3',              label: 'MiniMax M3 (opencode Go)',           tier: 'high',      cost: 'subscription' },
  { id: 'opencode-go:qwen3.7-plus',        provider: 'opencode-go',  model: 'qwen3.7-plus',            label: 'Qwen 3.7 Plus (opencode Go)',        tier: 'high',      cost: 'subscription' },
  { id: 'opencode-go:qwen3.7-max',         provider: 'opencode-go',  model: 'qwen3.7-max',             label: 'Qwen 3.7 Max (opencode Go)',         tier: 'high',      cost: 'subscription' },
  { id: 'opencode-go:kimi-k2.5',           provider: 'opencode-go',  model: 'kimi-k2.5',               label: 'Kimi K2.5 (opencode Go)',            tier: 'reasoning', cost: 'subscription' },
  { id: 'opencode-go:kimi-k2.6',           provider: 'opencode-go',  model: 'kimi-k2.6',               label: 'Kimi K2.6 (opencode Go)',            tier: 'reasoning', cost: 'subscription' },
];

export const TASKS = [
  { id: 'post_scoring',      label: 'Post Scoring',           complexity: 'low',       volume: 'high',     timeoutMs: 60000,
    description: 'Per-post pain-signal scoring during Reddit/Bluesky scout runs. High volume - runs 10 batches of 15 posts in parallel.',
    default: 'copilot:gpt-5-mini',
    defaultByProvider: {
      copilot: 'copilot:gpt-5-mini',
      'claude-cli': 'claude-cli:haiku',
      opencode: 'opencode-go:deepseek-v4-flash',
      sdk: 'sdk:haiku',
      gemini: 'gemini:2.5-flash',
    } },
  { id: 'query_suggestion',  label: 'Query Suggestion',       complexity: 'medium',    volume: 'one-time', timeoutMs: 60000,
    description: 'Generates 10 search URLs and angles when adding queries.',
    default: 'copilot:gpt-5-mini',
    defaultByProvider: {
      copilot: 'copilot:claude-haiku-4.5',
      'claude-cli': 'claude-cli:sonnet',
      opencode: 'opencode-go:mimo-v2.5-pro',
      sdk: 'sdk:haiku',
      gemini: 'gemini:2.5-flash',
    } },
  { id: 'query_refinement',  label: 'Query Refinement',       complexity: 'reasoning', volume: 'one-time', timeoutMs: 300000,
    description: 'Reviews current queries against report findings and recommends which to disable or add next. Uses a 5-minute timeout because report-backed refinement can run longer than basic suggestions.',
    default: 'claude-cli:opus',
    defaultByProvider: {
      copilot: 'copilot:claude-sonnet-4.6',
      'claude-cli': 'claude-cli:opus',
      opencode: 'opencode-go:kimi-k2.5',
      sdk: 'sdk:haiku',
      gemini: 'gemini:2.5-flash',
    } },
  { id: 'report_clustering', label: 'Research Report',        complexity: 'reasoning', volume: 'one-time', timeoutMs: 300000,
    description: 'Clusters posts into pain themes and proposes product angles. One-time per report; 5-minute timeout.',
    default: 'claude-cli:opus',
    defaultByProvider: {
      copilot: 'copilot:claude-sonnet-4.6',
      'claude-cli': 'claude-cli:opus',
      opencode: 'opencode-go:kimi-k2.5',
      sdk: 'sdk:haiku',
      gemini: 'gemini:2.5-flash',
    } },
  { id: 'draft_generation',  label: 'Draft Comment / DM',     complexity: 'medium',    volume: 'per-post', timeoutMs: 60000,
    description: 'Drafts contextual replies and direct messages for outreach.',
    default: 'claude-cli:haiku',
    defaultByProvider: {
      copilot: 'copilot:claude-sonnet-4.6',
      'claude-cli': 'claude-cli:sonnet',
      opencode: 'opencode-go:mimo-v2.5-pro',
      sdk: 'sdk:haiku',
      gemini: 'gemini:2.5-flash',
    } },
  { id: 'google_analysis',   label: 'Google Search Analysis', complexity: 'low',       volume: 'high',     timeoutMs: 60000,
    description: 'Per-result product extraction plus relevance and intent scoring in Google scout.',
    default: 'copilot:claude-haiku-4.5',
    defaultByProvider: {
      copilot: 'copilot:gpt-5-mini',
      'claude-cli': 'claude-cli:haiku',
      opencode: 'opencode-go:deepseek-v4-flash',
      sdk: 'sdk:haiku',
      gemini: 'gemini:2.5-flash',
    } },
  { id: 'advisor_council',   label: 'Advisor Council',         complexity: 'reasoning', volume: 'one-time', timeoutMs: 300000,
    description: 'Five-advisor council critique of a completed research report.',
    default: 'claude-cli:opus',
    defaultByProvider: {
      copilot: 'copilot:claude-sonnet-4.6',
      'claude-cli': 'claude-cli:opus',
      opencode: 'opencode-go:kimi-k2.5',
      sdk: 'sdk:haiku',
      gemini: 'gemini:2.5-flash',
    } },
];

const TIER_RANK = { low: 0, medium: 1, high: 2, reasoning: 3 };
const VALID_TIERS = new Set(Object.keys(TIER_RANK));

function getTierRank(tier, label) {
  if (!VALID_TIERS.has(tier)) {
    throw new RangeError(`Invalid ${label}: ${tier}`);
  }

  return TIER_RANK[tier];
}

export function isOverkill(modelTier, taskComplexity) {
  const modelRank = getTierRank(modelTier, 'modelTier');
  const taskRank = getTierRank(taskComplexity, 'taskComplexity');

  return modelRank > taskRank + 1;
}

export function isUnderkill(modelTier, taskComplexity) {
  const modelRank = getTierRank(modelTier, 'modelTier');
  const taskRank = getTierRank(taskComplexity, 'taskComplexity');

  return modelRank < taskRank;
}

export function getModel(id) {
  return MODEL_CATALOG.find(m => m.id === id);
}

export function getTask(id) {
  return TASKS.find(t => t.id === id);
}

/**
 * Returns the recommended model ID for a task given the user's provider.
 * Falls back to the task's hardcoded default when the provider is not in
 * defaultByProvider, and returns null when the task itself is not found.
 */
export function getDefaultForTask(taskId, provider) {
  const task = getTask(taskId);
  return task?.defaultByProvider?.[provider] || task?.default || null;
}

File: apps/api/internal/ai/catalog.go
package ai

// ModelEntry represents a single AI model available in the catalog.
type ModelEntry struct {
	ID       string `json:"id"`
	Provider string `json:"provider"`
	Model    string `json:"model"`
	Label    string `json:"label"`
	Tier     string `json:"tier"`
	Cost     string `json:"cost"`
}

// TaskEntry represents a task that can be configured with a model.
type TaskEntry struct {
	ID                string            `json:"id"`
	Label             string            `json:"label"`
	Complexity        string            `json:"complexity"`
	Volume            string            `json:"volume"`
	TimeoutMs         int               `json:"timeoutMs"`
	Description       string            `json:"description"`
	Default           string            `json:"default"`
	DefaultByProvider map[string]string `json:"defaultByProvider"`
}

// ModelCatalog is the full list of available AI models, mirroring catalog.js MODEL_CATALOG.
var ModelCatalog = []ModelEntry{
	{ID: "copilot:gpt-4.1", Provider: "copilot", Model: "gpt-4.1", Label: "GPT-4.1 (Copilot)", Tier: "medium", Cost: "usage-based"},
	{ID: "copilot:gpt-5-mini", Provider: "copilot", Model: "gpt-5-mini", Label: "GPT-5 mini (Copilot)", Tier: "low", Cost: "usage-based"},
	{ID: "copilot:gpt-5.4-mini", Provider: "copilot", Model: "gpt-5.4-mini", Label: "GPT-5.4 mini (Copilot)", Tier: "medium", Cost: "usage-based"},
	{ID: "copilot:gpt-5.4", Provider: "copilot", Model: "gpt-5.4", Label: "GPT-5.4 (Copilot)", Tier: "high", Cost: "usage-based"},
	{ID: "copilot:gpt-5.5", Provider: "copilot", Model: "gpt-5.5", Label: "GPT-5.5 (Copilot)", Tier: "reasoning", Cost: "usage-based"},
	{ID: "copilot:claude-haiku-4.5", Provider: "copilot", Model: "claude-haiku-4.5", Label: "Claude Haiku 4.5 (Copilot)", Tier: "low", Cost: "usage-based"},
	{ID: "copilot:claude-sonnet-4.6", Provider: "copilot", Model: "claude-sonnet-4.6", Label: "Claude Sonnet 4.6 (Copilot)", Tier: "medium", Cost: "usage-based"},
	{ID: "claude-cli:haiku", Provider: "claude-cli", Model: "haiku", Label: "Claude Haiku (CLI)", Tier: "low", Cost: "token-billed"},
	{ID: "claude-cli:sonnet", Provider: "claude-cli", Model: "sonnet", Label: "Claude Sonnet (CLI)", Tier: "medium", Cost: "token-billed"},
	{ID: "claude-cli:opus", Provider: "claude-cli", Model: "opus", Label: "Claude Opus (CLI)", Tier: "reasoning", Cost: "token-billed"},
	{ID: "sdk:haiku", Provider: "sdk", Model: "claude-haiku-4-5-20251001", Label: "Claude Haiku (SDK)", Tier: "low", Cost: "token-billed"},
	{ID: "gemini:2.5-flash", Provider: "gemini", Model: "gemini-2.5-flash", Label: "Gemini 2.5 Flash", Tier: "low", Cost: "token-billed"},
	{ID: "gemini:3.5-flash", Provider: "gemini", Model: "gemini-3.5-flash", Label: "Gemini 3.5 Flash", Tier: "reasoning", Cost: "token-billed"},
	{ID: "gemini:2.5-pro", Provider: "gemini", Model: "gemini-2.5-pro", Label: "Gemini 2.5 Pro", Tier: "medium", Cost: "token-billed"},
	{ID: "gemini:3.1-pro", Provider: "gemini", Model: "gemini-3.1-pro", Label: "Gemini 3.1 Pro", Tier: "high", Cost: "token-billed"},
	{ID: "opencode:big-pickle", Provider: "opencode", Model: "big-pickle", Label: "Big Pickle (opencode)", Tier: "medium", Cost: "free (0x)"},
	{ID: "opencode:deepseek-v4-flash-free", Provider: "opencode", Model: "deepseek-v4-flash-free", Label: "DeepSeek V4 Flash Free (opencode)", Tier: "medium", Cost: "free (0x)"},
	{ID: "opencode:mimo-v2.5-free", Provider: "opencode", Model: "mimo-v2.5-free", Label: "Mimo v2.5 Free (opencode)", Tier: "medium", Cost: "free (0x)"},
	{ID: "opencode:nemotron-3-ultra-free", Provider: "opencode", Model: "nemotron-3-ultra-free", Label: "Nemotron 3 Ultra Free (opencode)", Tier: "high", Cost: "free (0x)"},
	{ID: "opencode:north-mini-code-free", Provider: "opencode", Model: "north-mini-code-free", Label: "North Mini Code Free (opencode)", Tier: "medium", Cost: "free (0x)"},
	{ID: "opencode-go:deepseek-v4-flash", Provider: "opencode-go", Model: "deepseek-v4-flash", Label: "DeepSeek V4 Flash (opencode Go)", Tier: "medium", Cost: "subscription"},
	{ID: "opencode-go:mimo-v2.5", Provider: "opencode-go", Model: "mimo-v2.5", Label: "Mimo v2.5 (opencode Go)", Tier: "medium", Cost: "subscription"},
	{ID: "opencode-go:glm-5.1", Provider: "opencode-go", Model: "glm-5.1", Label: "GLM-5.1 (opencode Go)", Tier: "high", Cost: "subscription"},
	{ID: "opencode-go:glm-5.2", Provider: "opencode-go", Model: "glm-5.2", Label: "GLM-5.2 (opencode Go)", Tier: "reasoning", Cost: "subscription"},
	{ID: "opencode-go:mimo-v2.5-pro", Provider: "opencode-go", Model: "mimo-v2.5-pro", Label: "Mimo v2.5 Pro (opencode Go)", Tier: "high", Cost: "subscription"},
	{ID: "opencode-go:minimax-m2.7", Provider: "opencode-go", Model: "minimax-m2.7", Label: "MiniMax M2.7 (opencode Go)", Tier: "high", Cost: "subscription"},
	{ID: "opencode-go:qwen3.6-plus", Provider: "opencode-go", Model: "qwen3.6-plus", Label: "Qwen 3.6 Plus (opencode Go)", Tier: "medium", Cost: "subscription"},
	{ID: "opencode-go:deepseek-v4-pro", Provider: "opencode-go", Model: "deepseek-v4-pro", Label: "DeepSeek V4 Pro (opencode Go)", Tier: "high", Cost: "subscription"},
	{ID: "opencode-go:minimax-m3", Provider: "opencode-go", Model: "minimax-m3", Label: "MiniMax M3 (opencode Go)", Tier: "high", Cost: "subscription"},
	{ID: "opencode-go:qwen3.7-plus", Provider: "opencode-go", Model: "qwen3.7-plus", Label: "Qwen 3.7 Plus (opencode Go)", Tier: "high", Cost: "subscription"},
	{ID: "opencode-go:qwen3.7-max", Provider: "opencode-go", Model: "qwen3.7-max", Label: "Qwen 3.7 Max (opencode Go)", Tier: "high", Cost: "subscription"},
	{ID: "opencode-go:kimi-k2.6", Provider: "opencode-go", Model: "kimi-k2.6", Label: "Kimi K2.6 (opencode Go)", Tier: "high", Cost: "subscription"},
	{ID: "opencode-go:kimi-k2.7-code", Provider: "opencode-go", Model: "kimi-k2.7-code", Label: "Kimi K2.7 Code (opencode Go)", Tier: "high", Cost: "subscription"},
}

// Tasks is the full list of configurable tasks, mirroring catalog.js TASKS.
var Tasks = []TaskEntry{
	{
		ID: "post_scoring", Label: "Post Scoring", Complexity: "low", Volume: "high", TimeoutMs: 60000,
		Description: "Per-post pain-signal scoring during Reddit/Bluesky scout runs. High volume - runs 10 batches of 15 posts in parallel.",
		Default:     "copilot:gpt-5-mini",
		DefaultByProvider: map[string]string{
			"copilot":    "copilot:gpt-5-mini",
			"claude-cli": "claude-cli:haiku",
			"opencode":   "opencode-go:deepseek-v4-flash",
			"sdk":        "sdk:haiku",
			"gemini":     "gemini:2.5-flash",
		},
	},
	{
		ID: "query_suggestion", Label: "Query Suggestion", Complexity: "medium", Volume: "one-time", TimeoutMs: 60000,
		Description: "Generates 10 search URLs and angles when adding queries.",
		Default:     "copilot:gpt-5.4-mini",
		DefaultByProvider: map[string]string{
			"copilot":    "copilot:gpt-5.4-mini",
			"claude-cli": "claude-cli:sonnet",
			"opencode":   "opencode-go:mimo-v2.5-pro",
			"sdk":        "sdk:haiku",
			"gemini":     "gemini:3.1-pro",
		},
	},
	{
		ID: "query_refinement", Label: "Query Refinement", Complexity: "reasoning", Volume: "one-time", TimeoutMs: 300000,
		Description: "Reviews current queries against report findings and recommends which to disable or add next. Uses a 5-minute timeout because report-backed refinement can run longer than basic suggestions.",
		Default:     "claude-cli:opus",
		DefaultByProvider: map[string]string{
			"copilot":    "copilot:gpt-5.5",
			"claude-cli": "claude-cli:opus",
			"opencode":   "opencode-go:glm-5.2",
			"sdk":        "sdk:haiku",
			"gemini":     "gemini:3.5-flash",
		},
	},
	{
		ID: "report_clustering", Label: "Research Report", Complexity: "reasoning", Volume: "one-time", TimeoutMs: 300000,
		Description: "Clusters posts into pain themes and proposes product angles. One-time per report; 5-minute timeout.",
		Default:     "claude-cli:opus",
		DefaultByProvider: map[string]string{
			"copilot":    "copilot:gpt-5.5",
			"claude-cli": "claude-cli:opus",
			"opencode":   "opencode-go:glm-5.2",
			"sdk":        "sdk:haiku",
			"gemini":     "gemini:3.5-flash",
		},
	},
	{
		ID: "draft_generation", Label: "Draft Comment / DM", Complexity: "medium", Volume: "per-post", TimeoutMs: 60000,
		Description: "Drafts contextual replies and direct messages for outreach.",
		Default:     "claude-cli:sonnet",
		DefaultByProvider: map[string]string{
			"copilot":    "copilot:claude-sonnet-4.6",
			"claude-cli": "claude-cli:sonnet",
			"opencode":   "opencode-go:mimo-v2.5-pro",
			"sdk":        "sdk:haiku",
			"gemini":     "gemini:3.1-pro",
		},
	},
	{
		ID: "google_analysis", Label: "Google Search Analysis", Complexity: "low", Volume: "high", TimeoutMs: 60000,
		Description: "Per-result product extraction plus relevance and intent scoring in Google scout.",
		Default:     "copilot:claude-haiku-4.5",
		DefaultByProvider: map[string]string{
			"copilot":    "copilot:gpt-5-mini",
			"claude-cli": "claude-cli:haiku",
			"opencode":   "opencode-go:deepseek-v4-flash",
			"sdk":        "sdk:haiku",
			"gemini":     "gemini:2.5-flash",
		},
	},
	{
		ID: "advisor_council", Label: "Advisor Council", Complexity: "reasoning", Volume: "one-time", TimeoutMs: 300000,
		Description: "Five-advisor council critique of a completed research report.",
		Default:     "claude-cli:opus",
		DefaultByProvider: map[string]string{
			"copilot":    "copilot:gpt-5.5",
			"claude-cli": "claude-cli:opus",
			"opencode":   "opencode-go:glm-5.2",
			"sdk":        "sdk:haiku",
			"gemini":     "gemini:3.5-flash",
		},
	},
	{
		ID: "prompt_suggestion", Label: "Prompt Suggestion", Complexity: "medium", Volume: "one-time", TimeoutMs: 60000,
		Description: "Generates three complete prompt drafts for a given platform+type slot, seeded with project context and existing queries.",
		Default:     "copilot:gpt-5.4-mini",
		DefaultByProvider: map[string]string{
			"copilot":    "copilot:gpt-5.4-mini",
			"claude-cli": "claude-cli:sonnet",
			"opencode":   "opencode-go:mimo-v2.5-pro",
			"sdk":        "sdk:haiku",
			"gemini":     "gemini:3.1-pro",
		},
	},
}

// GetModel returns the model entry with the given ID, or nil if not found.
func GetModel(id string) *ModelEntry {
	for i := range ModelCatalog {
		if ModelCatalog[i].ID == id {
			return &ModelCatalog[i]
		}
	}
	return nil
}

// GetTask returns the task entry with the given ID, or nil if not found.
func GetTask(id string) *TaskEntry {
	for i := range Tasks {
		if Tasks[i].ID == id {
			return &Tasks[i]
		}
	}
	return nil
}
